### PR TITLE
Re-check for additional writer before writing new name key

### PR DESCRIPTION
**Pull request by palfrey**
Fixes the race condition from https://github.com/palfrey/serial_test/pull/45 and adds a stress test for it.

### CODE REVIEW

File: serial_test/src/code_lock.rs [hunk 1]
```diff
@@ -12,16 +12,18 @@ lazy_static! {
...
-            .unwrap()
-            .deref_mut()
-            .insert(name.to_string(), ReentrantMutex::new(()));
+        let lock = LOCKS.upgradable_read();
+        if lock.deref().contains_key(name) {
+            return; // another writer got here between unlock of the read and re-acquire
+        }
+        let mut upgraded_lock =
+            RwLockUpgradableReadGuard::<'_, HashMap<String, ReentrantMutex<()>>>::upgrade(lock);
+        upgraded_lock.insert(name.to_string(), ReentrantMutex::new(()));
```
Comment (FSMaxB) [inline serial_test/src/code_lock.rs:26]: Isn't this the same race condition once again? Imagine that both threads hold the `UpgradableReadGuard` and both check that the key is not contained at the same time. Then both race for the `upgrade` to a `WriteGuard` and the winner inserts the first Mutex, the looser replaces it with a different one.

**Review by FSMaxB — COMMENTED** [review]
I haven't really looked at the test yet, but I noticed that `cargo test` seems to deadlock for me with this change (in `test_with_key`)

File: serial_test/src/code_lock.rs (hunk 1 above)
Comment (palfrey) [reply serial_test/src/code_lock.rs:26]: That shouldn't happen (see https://docs.rs/lock_api/0.4.5/lock_api/struct.RwLock.html#method.upgradable_read "The calling thread will be blocked until there are no more writers or other upgradable reads which hold the lock") but I'm seeing bits of test locks as well which I'm still sifting through.

Comment (FSMaxB) [reply serial_test/src/code_lock.rs:26]: You're right, this should work because no two upgradable readers can hold the lock at the same time.

**Pull request merged**